### PR TITLE
ALC-183 gradual weight update data

### PR DESCRIPTION
**Pull request by hosainnet**
Introduce abi call integration to be able to extract configurable pool metadata, particularly gradual weight updates and the start/end time for the update 

Example usage: 

```
  const address = "0x8079aadf0514e6b2d6ffc19a5a42c95ee83cabae";

  const [data, setData] = useState<ConfigurableRightsPool | null>(null);

  useEffect(() => {
    fetchPoolMetadata(address).then((value) => {
      setData(value);
    });
  }, []);

  console.log(data);
```

- [x] Branch name corresponds to Linear ticket number.
- [x] Added sufficient integration and unit tests.
- [x] Addressed a single concern in the least number of changed lines as
      possible.
- [x] Only fixed/added the functionality in question **OR** address wide-spread
      whitespace/style issues, not both.

### CODE REVIEW

File: src/_copper/utils.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      ]),
+      options || {}
+    );
+    return response.map((call: BytesLike, i: any) =>
+      itf.decodeFunctionResult(calls[i][1], call)
+    );
+  } catch (e) {
+    return Promise.reject();
+  }
+}
```
Comment (hosainnet) [inline src/_copper/utils.ts:32]: Based on https://github.com/balancer-labs/pool-management-vue/blob/develop/src/_balancer/utils.ts#L10

File: src/constants/network/index.ts [hunk 2]
```diff
@@ -1,14 +1,19 @@
 import { ethers } from "ethers";
 
+const common = {
+  multicallAddress: "0xeefBa1e63905eF1D7ACbA5a8513c70307C1cE441",
```
Comment (hosainnet) [inline src/constants/network/index.ts:4]: https://github.com/balancer-labs/pool-management-vue/blob/d03939af8c9136be58e7c1b95fabfe0bb6c9c502/src/config/homestead.json#L17

File: src/helpers/abi/Multicall.json [hunk 3]
```diff
@@ -0,0 +1,146 @@
+{
```
Comment (hosainnet) [inline src/helpers/abi/Multicall.json:1]: https://github.com/balancer-labs/pool-management-vue/blob/master/src/helpers/abi/Multicall.json

File: src/helpers/crpMetadataFetcher.test.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
+import { BigNumber } from "ethers";
+import { fetchPoolMetadata } from "helpers/crpMetadataFetcher";
+import * as utils from "utils/date";
+import * as copperUtils from "../_copper/utils";
+import provider from "../helpers/provider";
+
```
Comment (hosainnet) [inline src/helpers/crpMetadataFetcher.test.ts:6]: Testing map of multicall to typed response

File: src/helpers/crpMetadataFetcher.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import { ConfigurableRightsPool } from "types/ConfigurableRightsPool";
+import { blockNumberToTimestamp } from "utils/date";
+import provider from "../helpers/provider";
+
+const crpAbi = require("helpers/abi/ConfigurableRightsPool.json");
+
+export async function fetchPoolMetadata(
+  address: String
+): Promise<ConfigurableRightsPool> {
+  return provider.getBlockNumber().then((currentBlock) => {
```
Comment (hosainnet) [inline src/helpers/crpMetadataFetcher.ts:12]: Largely based on https://github.com/balancer-labs/pool-management-vue/blob/develop/src/_balancer/pool.ts#L72

File: src/utils/date.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  currentBlockNumber: number,
+  blockNumber: number
+): number {
+  const AVG_BLOCK_TIMES = {
+    1: 13,
+    42: 5,
+  } as any;
+  const avgBlockTime = AVG_BLOCK_TIMES[constants.network.networkId as number];
+  return currentTime + avgBlockTime * 1000 * (blockNumber - currentBlockNumber);
+}
```
Comment (hosainnet) [inline src/utils/date.ts:14]: https://github.com/balancer-labs/pool-management-vue/blob/develop/src/helpers/utils.ts#L272

File: src/constants/network/index.ts [hunk 7]
```diff
@@ -1,14 +1,19 @@
...
+  rpcUrl: `https://mainnet.infura.io/v3/${process.env.REACT_APP_INFURA_API_KEY}`,
```
Comment (ledes) [inline src/constants/network/index.ts:5]: Do we need to add our own keys in a env file for working locallly?

File: src/services/configurable-pool-service/index.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        totalShares,
+        rights,
+        bspCap,
+        crpController,
+        minimumWeightChangeBlockPeriod,
+        addTokenTimeLockInBlocks,
+        { startBlock, endBlock },
+      ] = value;
+      return {
+        isPublicSwap: isPublicSwap[0],
```
Comment (ledes) [inline src/services/configurable-pool-service/index.ts:44]: why isPublicSwap an array?

File: src/services/configurable-pool-service/index.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        name: name.toString(),
+        symbol: symbol.toString(),
+        totalShares: formatUnits(totalShares.toString(), decimals),
+        rights: [],
+        bspCap: formatUnits(bspCap.toString(), decimals),
+        crpController: crpController[0],
+        // eslint-disable-next-line max-len
+        minimumWeightChangeBlockPeriod: minimumWeightChangeBlockPeriod.toString(),
```
Comment (ledes) [inline src/services/configurable-pool-service/index.ts:52]: you can probably get away without  the eslint disable line if you  create a variable for minimumWeightChangeBlockPeriod.toString() and use it here

File: src/utils/date.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+export function blockNumberToTimestamp(
+  currentTime: number,
+  currentBlockNumber: number,
+  blockNumber: number
+): number {
+  const AVG_BLOCK_TIMES = {
+    1: 13,
+    42: 5,
+  } as any;
```
Comment (ledes) [inline src/utils/date.ts:11]: can  you add a comment for why we use these numbers here?

**Review by ledes — COMMENTED** [review]
Looks really good!

File: src/utils/date.ts (hunk 10 above)
Comment (hosainnet) [reply src/utils/date.ts]: Moved to constants and added a comment 👍

File: src/services/configurable-pool-service/index.ts (hunk 9 above)
Comment (hosainnet) [reply src/services/configurable-pool-service/index.ts]: I ended up renaming minimumWeightChangeBlockPeriod to minWeightChangeBlockPeriod

File: src/constants/network/index.ts (hunk 7 above)
Comment (hosainnet) [reply src/constants/network/index.ts:5]: Yeah good point, I added an .env.local.example and updated README.md

File: src/services/configurable-pool-service/index.ts (hunk 8 above)
Comment (hosainnet) [reply src/services/configurable-pool-service/index.ts:44]: It's  just the way that multicall returns results, everything is an array even for single values

File: .env.local.example [hunk 11]
```diff
@@ -0,0 +1,3 @@
+NODE_PATH=src
```
Comment (pygeek) [inline .env.local.example:1]: Is there a reason you added NODE_PATH or REACT_APP_NETWORK to this file?

File: src/services/configurable-pool-service/utils.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+import { JsonRpcProvider } from "@ethersproject/providers";
+import { BytesLike } from "@ethersproject/bytes";
+
+const MULTICALL = constants.network.multicallAddress;
+
+export async function multicall(
+  provider: JsonRpcProvider,
+  abi: string,
+  calls: any[][],
+  options?: any
```
Comment (CryptoOsanu) [inline src/services/configurable-pool-service/utils.ts:14]: Do you think we could have a type or interface for the options parameter?

File: src/helpers/abi/Multicall.json (hunk 3 above)
Comment (pygeek) [reply src/helpers/abi/Multicall.json:1]: constants/blancer/abi/multicall.json

Comment (hosainnet) [reply src/helpers/abi/Multicall.json]: 👍

File: .env.local.example (hunk 11 above)
Comment (hosainnet) [reply .env.local.example]: Removed as they're defaulted 👍

File: src/services/configurable-pool-service/utils.ts (hunk 12 above)
Comment (hosainnet) [reply src/services/configurable-pool-service/utils.ts]: Removed as it is unused

**Review by ledes — APPROVED** [review]

**Pull request merged**